### DISCUSSION TITLE
自动将日志文件每天分1个文件，再自己压缩的方法

### DISCUSSION THREAD

**Discussion by chika0801** [XTLS/Xray-core#2369] [Ideas]
当你要保存日志到文件中时，一般是用下面的

```
    "log": {
        "access": "/var/log/xray/access.log",
        "error": "/var/log/xray/error.log",
        "loglevel": "warning",
        "dnsLog": false
    },
```

但是默认不会每天分1个文件，再自己压缩文件，询问了GTP，方法如下

在 /etc/logrotate.d 新建一个文件如 xray，内容如下，保存后就行了，每天会自动执行的，我在Debian10中测试过了。其它不懂的自己问GPT吧。

```
/var/log/xray/access.log {
    daily
    missingok
    rotate 30
    compress
    delaycompress
    notifempty
    create 0600 nobody nogroup
}

/var/log/xray/error.log {
    daily
    missingok
    rotate 30
    compress
    delaycompress
    notifempty
    create 0600 nobody nogroup
}
```

测试配置文件命令

```
logrotate -d /etc/logrotate.d/xray
```

比如 nginx 的日志自己会如下图

![屏幕截图 2023-07-25 004158](https://github.com/XTLS/Xray-core/assets/88967758/8854f791-79f4-4b3f-8e59-0d0d63334ce0)

**Comment by chise0713** [comment]
附上一个systemd的服务

```service

[Unit]
Description=Rotate log files
Documentation=man:logrotate(8)

[Service]
Type=oneshot
ExecStart=/usr/sbin/logrotate /etc/logrotate.d/%i

```

和一个systemd的计时器

```service

[Unit]
Description=Run logrotate for %i logs

[Timer]
OnCalendar=*-*-* 06:10:00 # 在每天早上 6:10 运行 logrotate for %i
Persistent=true

[Install]
WantedBy=timers.target

```

把需要logrotate的配置文件丢在 `/etc/logrotate.d/` 里就行。
使用 `systemctl enable --now logrotate@xray.timer` 激活计时器，计时器会在 `OnCalendar` 设置的时间运行 `logrotate@xray.service` 。
更多示例:
 `systemctl enable --now logrotate@v2ray.timer` 激活  `logrotate@v2ray.service` 
 `systemctl enable --now logrotate@nginx.timer` 激活  `logrotate@nginx.service` 
只需要把 `@` 后面的字符换成你在 `/etc/logrotate.d/` 配置文件的文件名就行

~~顺便~~

> 但是默认不会每天分1个文件，再自己压缩文件，询问了GTP，方法如下

~~的GPT打错啦！~~

**Reply by chika0801** [reply]
学习了

**Reply by chika0801** [reply]
大佬能发个 pr 到 安装脚本仓库吗 https://github.com/XTLS/Xray-install

😝

**Reply by chise0713** [reply]
> 大佬能发个 pr 到 安装脚本仓库吗 https://github.com/XTLS/Xray-install
> 
> 😝

logrotate不一定包含在默认安装内，还是算了🤪

**Reply by chika0801** [reply]
原来如此🤔😘

**Reply by chise0713** [reply]
> 原来如此🤔😘


不过我可以试试看加个函数，用$arg调用它，改为可选项，在没有检测到logrotate时提示用户安装，明天看看xray-install的代码，今天睡了🥰

**Reply by chika0801** [reply]
弄成了，还可以给v2fly core install pr，写1个pr得2个 :heart_eyes:

这日志自己滚动的事，我记得有人在群问过，虽然平时基本没人问，不过确实人有问过。。。

**Reply by chise0713** [reply]
> 弄成了，还可以给v2fly core install pr，写1个pr得2个 😍
> 
> 这日志自己滚动的事，我记得有人在群问过，虽然平时基本没人问，不过确实人有问过。。。


https://github.com/XTLS/Xray-install/pull/62
开了个PR，~~v2fly那边的就先搁着明天再说~~

**Comment by chika0801** [comment]
我用了2台VPS，分别是Debian10和12的系统，我只在 /etc/logrotate.d/ 目录下创建了一个文件名为 xray ，我观察了2天，日志被正确的每天分成1份了。

我参考了 nginx 的文件，现在是写的这样的内容

```
/var/log/xray/*.log {
    daily
    missingok
    rotate 30
    compress
    delaycompress
    notifempty
    create 0600 nobody nogroup
}
```

我没有测试ubuntu和其它系统，只采用在 /etc/logrotate.d/ 目录增加一个文件是否能正常工作。

我查看了你的 https://github.com/XTLS/Xray-install/pull/62 ，你考虑把 xray 文件内容写成 `/var/log/xray/*.log` *号的形式吗？

另外，我想建议将PR中 默认你给的是 ` rotate 30` 30天改小一些怎样了？比如就3或7天好了。可能时间默认给短一些，日志尽量保存短一点（作为加到安装脚本里的参数值的话）

**Reply by chise0713** [reply]
> 我查看了你的 [XTLS/Xray-install#62](https://github.com/XTLS/Xray-install/pull/62) ，你考虑把 xray 文件内容写成 `/var/log/xray/*.log` *号的形式吗？
> 
> 另外，我想建议将PR中 默认你给的是 ` rotate 30` 30天改小一些怎样了？比如就3或7天好了。可能时间默认给短一些，日志尽量保存短一点（作为加到安装脚本里的参数值的话）

感谢，这就去改改


> 我没有测试ubuntu和其它系统，只采用在 /etc/logrotate.d/ 目录增加一个文件是否能正常工作。

不太明白这句话的意思，我PR内默认是如果没有这个目录自动创建目录之后自动创建配置文件。

**Comment by chika0801** [comment]
> 不太明白这句话的意思，我PR内默认是如果没有这个目录自动创建目录之后自动创建配置文件。

我是用debian10 12系统，DD完系统，自己就有  /etc/logrotate.d/ 目录。我是这个意思。其它系统我不清楚了。

**Reply by chise0713** [reply]
就和我之前说的logrotate不一定包含在一些系统的默认安装一样，比如说我使用的arch linux就没有这个文件夹，即使之后安装了logrotate也不会自动创建，所以就有了脚本的那一段。

**Comment by chika0801** [comment]
原来这样明白了
